vllm/v1/guided_decoding/__init__.py

```diff
@@ -0,0 +1,195 @@
+from __future__ import annotations
+
+import copy
+import enum
+import threading
+from concurrent.futures import ThreadPoolExecutor
+from dataclasses import dataclass, field
+from typing import TYPE_CHECKING, Dict, List, Optional, Set, Tuple
+
+import torch
+import xgrammar as xgr
+
+from vllm.config import VllmConfig
+from vllm.logger import init_logger
+from vllm.transformers_utils.tokenizer_group import init_tokenizer_from_configs
+
+if TYPE_CHECKING:
+    from vllm.v1.request import Request
+
+logger = init_logger(__name__)
+
+
+class GuidedDecodingOptions(enum.Enum):
+    json = enum.auto()
+    json_object = enum.auto()
+    regex = enum.auto()
+    grammar = enum.auto()
+    choice = enum.auto()
+
+
+GuidedDecodingKey = Tuple[GuidedDecodingOptions, str]
+MAX_ROLLBACK_TOKENS = 100
+
+
+def apply_bitmask(logits: torch.Tensor, vocab_mask: torch.Tensor,
+                  indices: List[int]) -> None:
+    xgr.apply_token_bitmask_inplace(logits, vocab_mask, indices=indices)
+
+
+@dataclass(slots=True, unsafe_hash=True)
+class Grammar:
+    # NOTE: This would be a generic-enough class for
+    # supporting different backends, in the future.
+    # For now, just xgrammar.
+    #
+    # TODO: support max_rollback_tokens
+    # https://xgrammar.mlc.ai/docs/api/python/index.html#xgrammar.GrammarMatcher.find_jump_forward_string
+    # for jump-forward decoding
+
+    vocab_size: int
+    matcher: xgr.GrammarMatcher = field(hash=False)
+    ctx: xgr.CompiledGrammar = field(hash=False)
+    max_rollback_tokens: int = field(default=MAX_ROLLBACK_TOKENS, kw_only=True)
+    num_processed_tokens: int = field(
+        default_factory=lambda: 0,
+        repr=False,
+        hash=False,
+        init=False,
+    )
+    _matcher_lock: threading.Lock = field(
+        default_factory=lambda: threading.Lock(),
+        repr=False,
+        init=False,
+        hash=False,
+    )
+
+    def accept_token(self, token: int) -> bool:
+        # NOTE: accept_token will determines whether we accept this token
+        # and will also update the machine state
+        with self._matcher_lock:
+            self.num_processed_tokens += 1
+            return self.matcher.accept_token(token)
+
+    # this should be ran in parallel with model decoding
+    def fill_bitmask(self, bitmask: torch.Tensor, idx: int) -> bool:
+        return self.matcher.fill_next_token_bitmask(bitmask, idx)
+
+    def rollback(self, num_tokens: int):
+        with self._matcher_lock:
+            if self.num_processed_tokens > 0:
+                self.num_processed_tokens -= num_tokens
+                self.matcher.rollback(num_tokens)
+
+    def reset(self):
+        self.num_processed_tokens = 0
+        self.matcher.reset()
+
+    def __copy__(self):
+        return Grammar(matcher=xgr.GrammarMatcher(self.ctx),
+                       vocab_size=self.vocab_size,
+                       ctx=self.ctx,
+                       max_rollback_tokens=self.max_rollback_tokens)
```

Comment:
I think we can simplify code in this file if we only cache the compiled grammar from xgrammar and not an instance of this `Grammar` class. It's the compiled grammar that's the only thing that gets reused. That way we'd just instantiate a new `Grammar` each time instead of doing a `copy.copy()` on a `Grammar` from the cache.

Reply:
I see. I think that the overhead of class is pretty minimal in this case.

The reasoning behind the grammar class is that from other class POV they don't have to worry too much about the internals (in case when we move to diff backends, it might be different context/concepts there, so hopefully this class is general enough such that if we need to do any copy, `copy.copy(...)` should be sufficient in most cases.)